api/grpcserver/v2alpha1/layer.go

```diff
@@ -0,0 +1,361 @@
+	"bytes"
+	"context"
+	"errors"
+	"io"
+
+	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
+	spacemeshv2alpha1 "github.com/spacemeshos/api/release/go/spacemesh/v2alpha1"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/events"
+	"github.com/spacemeshos/go-spacemesh/sql"
+	"github.com/spacemeshos/go-spacemesh/sql/builder"
+	"github.com/spacemeshos/go-spacemesh/sql/layers"
+)
+
+const (
+	Layer       = "layer_v2alpha1"
+	LayerStream = "layer_stream_v2alpha1"
+)
+
+func NewLayerStreamService(db sql.Executor) *LayerStreamService {
+	return &LayerStreamService{db: db}
+}
+
+type LayerStreamService struct {
+	db sql.Executor
+}
+
+func (s *LayerStreamService) RegisterService(server *grpc.Server) {
+	spacemeshv2alpha1.RegisterLayerStreamServiceServer(server, s)
+}
+
+func (s *LayerStreamService) RegisterHandlerService(mux *runtime.ServeMux) error {
+	return spacemeshv2alpha1.RegisterLayerStreamServiceHandlerServer(context.Background(), mux, s)
+}
+
+func (s *LayerStreamService) Stream(
+	request *spacemeshv2alpha1.LayerStreamRequest,
+	stream spacemeshv2alpha1.LayerStreamService_StreamServer,
+) error {
+	ctx := stream.Context()
+	var sub *events.BufferedSubscription[events.LayerUpdate]
+	if request.Watch {
+		matcher := layersMatcher{request, ctx}
+		var err error
+		sub, err = events.SubscribeMatched(matcher.match)
+		if err != nil {
+			return status.Error(codes.Internal, err.Error())
+		}
+		defer sub.Close()
+		if err := stream.SendHeader(metadata.MD{}); err != nil {
+			return status.Errorf(codes.Unavailable, "can't send header")
+		}
+	}
+
+	dbChan := make(chan *spacemeshv2alpha1.Layer, 100)
+	errChan := make(chan error, 1)
+
+	ops, err := toLayerOperations(toLayerRequest(request))
+	if err != nil {
+		return status.Error(codes.InvalidArgument, err.Error())
+	}
+	// send db data to chan to avoid buffer overflow
+	go func() {
+		defer close(dbChan)
+		if err := layers.IterateLayersWithBlockOps(s.db, ops, func(layer *layers.Layer) bool {
+			select {
+			case dbChan <- &spacemeshv2alpha1.Layer{Versioned: &spacemeshv2alpha1.Layer_V1{
+				V1: toLayer(layer),
+			}}:
+				return true
+			case <-ctx.Done():
+				// exit if the stream context is canceled
+				return false
+			}
+		}); err != nil {
+			errChan <- status.Error(codes.Internal, err.Error())
+			return
+		}
+	}()
+
+	var eventsOut <-chan events.LayerUpdate
+	var eventsFull <-chan struct{}
+	if sub != nil {
+		eventsOut = sub.Out()
+		eventsFull = sub.Full()
+	}
+
+	for {
+		select {
+		case rst := <-eventsOut:
+			var derr error
+			ops := builder.Operations{}
+			ops.Filter = append(ops.Filter, builder.Op{
+				Prefix: "l.",
+				Field:  builder.Id,
+				Token:  builder.Eq,
+				Value:  int64(rst.LayerID.Uint32()),
+			})
+			if err := layers.IterateLayersWithBlockOps(s.db, ops, func(layer *layers.Layer) bool {
```

Comment:
If I understand correctly, you expect to get only 1 layer back (`l.id == rs.LayerID`). How about adding a `Get(sql.Executor, types.LayerID) (layers.Layer, error)`  function in `layers` package?